h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/MatrixLoader.java

```diff
@@ -43,19 +45,23 @@ public final DMatrix get() throws XGBoostError {
         }
 
         @Override
-        public void writeExternal(ObjectOutput out) throws IOException {
-            out.writeLong(actualRows);
-            out.writeObject(response);
-            out.writeObject(weights);
-            out.writeObject(offsets);
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (!(o instanceof DMatrixProvider)) return false;
+            DMatrixProvider that = (DMatrixProvider) o;
+            return actualRows == that.actualRows &&
+                Arrays.equals(response, that.response) &&
+                Arrays.equals(weights, that.weights) &&
+                Arrays.equals(offsets, that.offsets);
         }
 
         @Override
-        public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
-            actualRows = in.readLong();
-            response = (float[]) in.readObject();
-            weights = (float[]) in.readObject();
-            offsets = (float[]) in.readObject();
+        public int hashCode() {
+            int result = Objects.hash(actualRows);
```

Comment:
This is a very expensive implementation of hashcode, number of rows + optionally maybe first few elements of the array should be good enough